mimic/rest/ironic_api.py

```diff
@@ -21,6 +21,20 @@ def __init__(self, core):
         """
         self.core = core
 
+    @app.route('/nodes', methods=['POST'])
+    def create_node(self, request):
+        """
+        Responds with response code 201 and returns the newly created node.
+        """
+        return self.core.ironic_node_store.create_node(request)
+
+    @app.route('/nodes/<string:node_id>', methods=['DELETE'])
+    def delete_node(self, request, node_id):
+        """
+        Responds with response code 204 and delete the node.
+        """
+        return self.core.ironic_node_store.delete_node(request, node_id)
```

Comment:
Perhaps we could list out the known use cases for modeling API behaviors
and see if we can improve on any of them individually before jumping to a
general solution.

For instance, allowing the backend to be down might not have to be specific
to the contents of each backend. Maybe instead Mimic keeps an up/down state
for each service and serves a generic 503 if it's down. Truncating the http
response in the middle could be done using middleware.

I don't expect to solve any new problems in this thread, but it seems to me
that 503s or truncated requests are concerns at the layer of HTTP or lower.
As a plugin author I would rather work and think in the domain of my
application, i.e., HTTP layer + 1. The simplest way I can think of to not
have this layer violation is to do HTTP-related stuff in the "view" and
only business stuff in the data stores, since the view already does HTTP
stuff. There are other possibilities for which we should start an issue or
discuss offline.

On Friday, September 18, 2015, Glyph <notifications@github.com> wrote:

> In mimic/rest/ironic_api.py
> <https://github.com/rackerlabs/mimic/pull/387#discussion_r39910535>:
>
> > @@ -21,6 +21,20 @@ def __init__(self, core):
> >          """
> >          self.core = core
> >
> > +    @app.route('/nodes', methods=['POST'])
> > +    def create_node(self, request):
> > +        """
> > +        Responds with response code 201 and returns the newly created node.
> > +        """
> > +        return self.core.ironic_node_store.create_node(request)
> > +
> > +    @app.route('/nodes/<string:node_id>', methods=['DELETE'])
> > +    def delete_node(self, request, node_id):
> > +        """
> > +        Responds with response code 204 and delete the node.
> > +        """
> > +        return self.core.ironic_node_store.delete_node(request, node_id)
>
> I agree that it *seems* like there's a layering violation here - and
> honestly, maybe there is one and we could figure out a better structure -
> but Mimic is a bit of a special case in terms of the correspondence between
> its "model" later and its "view" (or public-facing HTTP API layer). Mimic's
> models are modeling not just the abstract state of "servers" or "load
> balancers" or whatever, but *API behaviors* as well, which means that in
> the "model" layer we occasionally need representations of things like "the
> backend is down, return a 503 for some reason" or "randomly truncate the
> HTTP response in the middle, we saw that happen when talking to the load
> balancer API one time when trying to manipulate this one messed up CLB". As
> such, passing along the full HTTP request and all its capabilities seems
> like the best option available, although I'd love to hear other thoughts,
> because this definitely feels not quite right based on every *other*
> project I've ever worked on :).
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/rackerlabs/mimic/pull/387/files#r39910535>.
>